pallets/liquid-staking-two/src/lib.rs

```diff
@@ -0,0 +1,820 @@
+
+    /// Current era index on Relaychain.
+    ///
+    /// CurrentEra: EraIndex
+    #[pallet::storage]
+    #[pallet::getter(fn current_era)]
+    pub type CurrentEra<T: Config> = StorageValue<_, EraIndex, ValueQuery>;
+
+    /// Store operations and corresponding status during each era
+    ///
+    /// The operation include: bond/bond_extra/unbond/rebond/on_new_era/transfer_to_relaychain/record_reward/record_slash
+    /// The status include: start/processing/successed/failed
+    #[pallet::storage]
+    #[pallet::getter(fn staking_operation_history)]
+    pub type StakingOperationHistory<T: Config> = StorageDoubleMap<
+        _,
+        Blake2_128Concat,
+        EraIndex,
+        Blake2_128Concat,
+        StakingOperationType,
+        Operation<T::BlockNumber>,
+    >;
+
+    // #[pallet::genesis_config]
+    // pub struct GenesisConfig {
+    //     pub exchange_rate: Rate,
+    //     pub reserve_factor: Ratio,
+    // }
+
+    // #[cfg(feature = "std")]
+    // impl Default for GenesisConfig {
+    //     fn default() -> Self {
+    //         Self {
+    //             exchange_rate: Rate::default(),
+    //             reserve_factor: Ratio::default(),
+    //         }
+    //     }
+    // }
+
+    // #[pallet::genesis_build]
+    // impl<T: Config> GenesisBuild<T> for GenesisConfig {
+    //     fn build(&self) {
+    //         ExchangeRate::<T>::put(self.exchange_rate);
+    //         ReserveFactor::<T>::put(self.reserve_factor);
+    //     }
+    // }
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T>
+    where
+        [u8; 32]: From<<T as frame_system::Config>::AccountId>,
+    {
+        #[pallet::weight(10_000)]
+        #[transactional]
+        pub fn trigger_new_era(
+            origin: OriginFor<T>,
+            era_index: EraIndex,
+        ) -> DispatchResultWithPostInfo {
+            // T::WithdrawOrigin::ensure_origin(origin)?;
+            let _who = ensure_signed(origin)?;
+            let current_era = Self::current_era();
+            PreviousEra::<T>::put(current_era);
+            CurrentEra::<T>::put(era_index);
+            Ok(().into())
+        }
+
+        //todo，record reward on each era, invoked by stake-client
+        // StakingPool = StakingPool + reward amount
+        // StakingPool/T::currency::total_issuance(liquidcurrency)
+        #[pallet::weight(10_000)]
+        #[transactional]
+        pub fn record_reward(
+            origin: OriginFor<T>,
+            agent: T::AccountId,
+            era_index: EraIndex,
+            #[pallet::compact] amount: Balance,
+        ) -> DispatchResultWithPostInfo {
+            // T::WithdrawOrigin::ensure_origin(origin)?;
+            let _who = ensure_signed(origin)?;
+            // ensure!(T::Members::contains(&agent), Error::<T>::IllegalAgent);
+            StakingPool::<T>::try_mutate(|m| -> DispatchResult {
```

Comment:
we should check if this agent has already one successful reward in this era, if yes then refuse the new one

Reply:
maybe the staking pool mutation should be done after this check

Reply:
Yeah, it mutates only if the status is `Ready`.